itemlist.lua

```diff
@@ -0,0 +1,231 @@
+		end
+	until s == nil
+
+	-- Find node on_receive_fields
+	local ndef = minetest.registered_nodes[minetest.get_node(pos).name]
+
+	if ndef and ndef.on_receive_fields then
+		update_context(player, { on_receive_fields = ndef.on_receive_fields } )
+	end
+	update_context(player, { node_pos = pos } )
+
+	-- Show formspec
+	minetest.show_formspec(playername, modname..':context_formspec', fs)
+end
+
+minetest.register_on_player_receive_fields(function(player, formname, fields)
+	if formname == modname..':context_formspec' then
+		local context = get_context(player)
+		if context == nil then return end
+
+		if context.on_receive_fields then
+			context.on_receive_fields(context.node_pos, '', fields, player)
+		end
+	end
+end)
+
+-- Specific functions
+
+local function get_nb_of_pages(context)
+	return math.ceil((#context.items + 1) / (style.lines * style.cols))
+end
+
+local function item_list_prepare(item_list)
+	local list = {}
+	for _, name in ipairs(item_list) do
+		local ndef = minetest.registered_items[name]
+		if ndef then
+			list[#list+1] = ndef
+		end
+	end
+	table.sort(list, function(a,b) return a.description < b.description end)
+	return list
+end
+
+local function show_fs(player)
+	local context = get_context(player)
+	if context == nil then return end
+
+	local fs = 'size['..(style.colsize * style.cols)..','
+		..(style.linesize * style.lines + 1.3)..']'
+		..default.gui_bg..default.gui_bg_img..default.gui_slots
+		..'label[1,0;'..context.title..']'
+		..'button_exit['..(style.colsize * style.cols - 2)..','
+		..(style.linesize * style.lines + 0.7)..';2,1;close;Close]'
+
+	if context.page > 1 then
+	 	fs = fs..'button['..(style.colsize * style.cols - 2)..',-0.2;1,1;prev;<]'
+	end
+
+	if context.page < get_nb_of_pages(context) then
+		fs = fs..'button['..(style.colsize * style.cols - 1)..',-0.2;1,1;next;>]'
+	end
+
+	local index = (style.lines * style.cols) * (context.page - 1)
+
+	for col = 1, style.cols do
+		local x = (col - 1) * style.colsize
+		for line = 1, style.lines do
+			index = index + 1
+			local y = line * style.linesize - 0.2
+			if context.items[index] then
+				fs = fs..'item_image['..x..','..y..';1,1;'
+				..context.items[index].name..']label['..(x+1)..','..(y+0.2)..';'
+				..(context.items[index].description or context.items[index].name)
+				..']'
+			end
+		end
+	end
+	minetest.show_formspec(context.playername, modname..':item_list', fs)
+end
+
+minetest.register_on_player_receive_fields(function(player, formname, fields)
+	if formname == modname..':item_list' then
+		local context = get_context(player)
+		if context == nil then return end
+
+		if fields.next then
+			context.page = context.page + 1
+			show_fs(context.playername)
+		end
+		if fields.prev then
+			context.page = context.page - 1
+			show_fs(context.playername)
+		end
+		if fields.quit == 'true' then
+			if context.node_pos then
+				-- Using after to avoid the "double close" bug
+				minetest.after(0, show_node_formspec, player, context.node_pos)
```

Comment:
Floating player reference. It may turn invalid any time after the current function is done executing. Please pass the player name instead of the player object reference.

Reply:
Done in 1c132d84b0bd4a862c6d29f73a83d6c1ee372259